dds/DCPS/DataReaderImpl.h

```diff
@@ -467,20 +467,50 @@ class OpenDDS_Dcps_Export DataReaderImpl
   {
   public:
     OwnershipManagerPtr(DataReaderImpl* reader)
-      : participant_( reader->is_exclusive_ownership_ ? reader->participant_servant_.lock() : RcHandle<DomainParticipantImpl>())
+      : participant_( (reader && reader->is_exclusive_ownership_) ? reader->participant_servant_.lock() : RcHandle<DomainParticipantImpl>())
     {
     }
     operator bool() const { return participant_.in(); }
     OwnershipManager* operator->() const
     {
-      return participant_->ownership_manager();
+      return participant_ ? participant_->ownership_manager() : 0;
     }
 
   private:
     RcHandle<DomainParticipantImpl> participant_;
   };
   friend class OwnershipManagerPtr;
 
+  struct OwnershipManagerScopedAccess {
+    OwnershipManagerScopedAccess() : om_(0), lock_result_(0) {}
+    OwnershipManagerScopedAccess(DataReaderImpl::OwnershipManagerPtr om) : om_(om), lock_result_(om_ ? om_->instance_lock_acquire() : 0) {}
+    ~OwnershipManagerScopedAccess() { release(); }
+
+    int release()
+    {
+      int result = 0;
+      if (om_ && !lock_result_) {
+        result = om_->instance_lock_release();
+      }
+      om_ = 0;
+      lock_result_ = 0;
+      return result;
+    }
+
+    void swap(OwnershipManagerScopedAccess& sa)
+    {
+      OwnershipManagerPtr om = om_;
+      om_ = sa.om_;
+      sa.om_ = om;
+      int lock_result = lock_result_;
+      lock_result_ = sa.lock_result_;
+      sa.lock_result_ = lock_result;
```

Comment:
Could each of these sets of 3 statements be replaced by std::swap? Could the whole method be replaced by std::swap of the OwnershipManagerScopedAccess?